### ISSUE TITLE
Renderer Rework: Initial Merge Tracking Issue

### ISSUE THREAD

**Issue by cart** [bevyengine/bevy#2535]
The Bevy Renderer Rework is starting to stabilize and it is time to start planning how to upstream it! The intent of this issue is to track the work required to merge the renderer rework as soon as possible. This isn't a "renderer feature wishlist". But feel free to discuss what you think should be on this list!

This is an issue to track work that still needs to be done. If there is a name in parentheses, that work has been claimed by someone.

For some history on this effort, check out:

* #2265
* #2351
* experimental pull requests in @cart's repo: https://github.com/cart/bevy/pulls?q=is%3Apr

Here is a [list of open prs against the pipelined-rendering branch](https://github.com/bevyengine/bevy/pulls?q=is%3Apr+is%3Aopen+base%3Apipelined-rendering)

## Missing Required Features

The new renderer must have (approximate) feature parity with the old renderer.

- [ ] actual pipelining
    * currently ownership in the new renderer is set up in a way to facilitate this, but _actual_ parallel pipelining in the context of winit will require some additional work
- [x] simple custom shaders (@cart)
    - [x] [Pipeline Specialization](#3031) (@cart) 
    - [x] [shaders-as-assets](#3031) (@cart)
    - [x] [hot shader reloading](#3031) (@cart)
    - [x] [custom materials](#2831) (@cart)
    - [x] [pipeline logic reuse](#2831) (@cart)
- [x] #2653 (@Davier)
    - [x] text (@StarArawn)
- [x] #2560 (@StarArawn)
- [x] #2537 (@superdump)
    - [x] re-enable gltf tangent imports #2741 (@superdump)
- [x] crevice bugfixing
    - [x] https://github.com/LPGhatguy/crevice/issues/29
    - [ ] investigate potential array alignment corner case (@cart)
- [x] #2543 (@superdump)
- [x] #2552  (@cart)
- [x] #2695 (@zicklag)
    - [x] #2717 (@cart) 
- [x] #3042  ~~#2541~~ (@superdump @cart)
- [x] Update BufferVec to use Queue (like UniformVec does) (#2847)
- [ ] Better `run_sub_graph()` input api
    * it shouldn't require correctly-ordered inputs. accept a "named" map instead
- [x] #2741(@superdump)
- [x] #2631 (@zicklag)
- [x] #2700 (@superdump)
- [x] #2555 (@cart)
- [x] [Add Opaque 3d render phase, sort that front-to-back, and sort transparent phase back-to-front](#3072) (@superdump)

## Missing Nice-To-Have Features

These aren't required for a merge, but would be very nice to have.

- [x] Visibility culling (probably requires merging and implementing https://github.com/bevyengine/rfcs/pull/12) (#2861)
    * This is an important part of the renderer api that isn't fleshed out yet and might meaningfully change the architecture.
- [x] #3060 ~~#2642~~ (@cart @zicklag)
- [x] #2726  (@superdump)
- [x] #2700  (@superdump)
- [ ] #2876 (@ChangeCaps)
- [x] #3153 (@superdump)
- [ ] shadow filtering (@superdump)
- [ ] make querying for "inner lifetimes" more ergonomic
- [ ] DrawFunction ergonomics (@cart)
- [ ] "high level" data binding (@cart)
- [x] [shader defs](#3031) (@cart)
- [x] #3137 (@cart)
- [x] ~~[depth pre-pass](#3072)~~ (@superdump)
  * Tabled for now because it generally reduces performance in our tests. In the future new cases like using the depth prepass for early SSAO might justify adding it.
- [ ] nicer api for flexible/custom vertex attributes (@cart)
- [ ] Generic `MaterialPlugin<T: Material>` for low boilerplate custom shaders  (@cart)
- [x] #3193 

## Discussions To Have Before Merging

- [ ] Consider RenderDevice + RenderQueue -> GpuDevice + GpuQueue
- [ ] Consider using atomic counter instead of UUID for render resource ids
- [ ] Should we keep the BevyDefault trait?
- [ ] Agree on the final "shadow enable/disable" api

## Steps to Merge

- [x] Move work from https://github.com/cart/bevy/tree/pipelined-rendering to https://github.com/bevyengine/bevy/tree/pipelined-rendering
- [x] Merge as many small, orthogonal, easy-to-retain-attribution changes from the pipelined-rendering branch as possible:
    - [x] AppBuilder -> App: #2531 (@bjorn3)
    - [x] SubApps (@cart)
    - [x] Break out SystemParam state and world lifetimes #2605 (@cart)
    - [x] ["Lifetimeless" system params](#2831) (@cart)
    - [x] #2283 (@cart)
    - [x] #2673 (@cart)
        - [x] world entity clearing (@cart)
        - [x] "invalid archetype" (@cart)
        - [x] stage configuration to disable automatic buffer application (@cart)
- [x] Decide whether or not to "vendor" crevice (we will)
- [ ] RFC for new renderer architecture (@cart)
- [x] Swap out old renderer crates for new renderer crates (after Missing Required Features section is complete)
    - [x] Port examples
    - [x] update docs
- [x] Manually merge (to prevent bors from squashing history and making attribution difficult)

**Comment by superdump** [comment]
Here's the PR for bevy_gltf2: https://github.com/bevyengine/bevy/pull/2537

**Comment by superdump** [comment]
I’ve got MSAA working but it’s hacky because we need one of:
* the ability to configure what node slots are available dynamically
* the ability to leave node slots disconnected and make them optional, doing a sensible automatic fallback where it makes sense
* hacks like connecting something to a node slot for the sake of filling the slot and passing a run sub graph input that isn’t used

**Comment by superdump** [comment]
Here's a work-in-progress PR for MSAA that I would appreciate some guidance on: https://github.com/bevyengine/bevy/pull/2541

I was trying to support dynamically switching MSAA sample count at run time, which is why it looks more complicated than it perhaps needs to be. I hit the wall mentioned in my last message when switching between 1 sample and multiple samples or vice versa.

However, I wanted to see if I could at least support switching between 2/4/8, but it turned out my Radeon Pro 5500M only supports 1 sample or 4x MSAA samples, so I couldn't test that.

I thought that as I had done most of the work for dynamically adjusting the MSAA sample count, I'd leave it in the PR and see what feedback I got.

**Comment by superdump** [comment]
And a small PR to log the adapter and wgpu backend being used at initialisation time. It's useful to see. https://github.com/bevyengine/bevy/pull/2542

**Comment by superdump** [comment]
I'll do `ClearColor` as well when there is feedback on MSAA as it's a small change and that touches the same code.

**Comment by superdump** [comment]
Infinite reverse right-handed PR here: https://github.com/bevyengine/bevy/pull/2543

It has a known bug of all geometry being black when looking along -Z that I am still debugging. Help welcome.

**Comment by superdump** [comment]
>     * Update BufferVec to use Queue (like UniformVec does)

What does this mean? I don't see anything about `UniformVec` that uses a `Queue` type.

**Comment by cart** [comment]
Great work @superdump. I'll dig in tomorrow / answer questions. Just a heads up that CI on `pipelined-rendering` is broken. Almost done fixing it in #2538

**Comment by superdump** [comment]
>     * the ability to configure what node slots are available dynamically

I added support for removing nodes, node edges, slot edges, and subgraphs. I've only tested use of the APIs for removal of slot edges but now MSAA can be toggled on the fly and I added that into the msaa_pipelined example - press 'm' to try it out.

**Comment by aevyrie** [comment]
Is there anything I can do to help move the primitive shapes RFC forward? I'd be happy to contribute towards implementation PRs once you've reviewed it.

**Comment by cart** [comment]
@aevyrie haha sadly not really (other than continuing to ping me so I don't forget and trying to get more reviews on it). I'll try to get to it in the next day or so as we do want to get the ball rolling on visibility.

**Comment by aevyrie** [comment]
Sounds good 🙂. I'd like to help with visibility/culling if you don't mind including me in those discussions.

**Comment by cart** [comment]
> I'd like to help with visibility/culling if you don't mind including me in those discussions.

I'd love to include you in those discussions :)

**Comment by Draghi** [comment]
Hi there, fresh face, just wanted to put my hand up for some of the writing/updating documentation and porting examples - if you're looking for helpers on that

**Comment by superdump** [comment]
> Hi there, fresh face, just wanted to volunteer for some of the writing/updating documentation and porting examples - if you're looking for helpers on that

Just dive in! :) Mention here what you’re working on and go for it.

**Comment by notverymoe** [comment]
Alright, I've made a small start just going through the changes. [Next I'll begin porting the examples](/notverymoe/bevy/tree/pipelined-rendering-examples), probably a good way to get familiar.

**Comment by cart** [comment]
Just a heads up that most of the examples wont need porting once we remove the old renderer and replace it with the new one. I don't think we should migrate examples until we've removed the old renderer (which we should wait to do until we have feature parity with the old renderer).

**Comment by cart** [comment]
@superdump 
> * Update BufferVec to use Queue (like UniformVec does)

Oops! I understand why this is confusing: I haven't merged those changes yet. They're available on my custom-shaders branch though: https://github.com/cart/bevy/blob/custom-shaders/pipelined/bevy_render2/src/render_resource/uniform_vec.rs. Sorry for the confusion. I crossed some wires :smile:

**Comment by kirawi** [comment]
Will this be in 0.6?

**Comment by cart** [comment]
That is the goal!

**Comment by Davier** [comment]
I'm starting to work on porting `bevy_ui`

**Comment by zicklag** [comment]
Just ran into the need to set the clear color with the new renderer. Would it be helpful if I added a way to do that? Also, would we want to use the same, `ClearColor` resource strategy as we did with the old renderer?

**Comment by Weibye** [comment]
This should have the label `C-Tracking-Issue`

**Comment by cart** [comment]
> Just ran into the need to set the clear color with the new renderer. Would it be helpful if I added a way to do that? Also, would we want to use the same, ClearColor resource strategy as we did with the old renderer?
> I'll take that! smile

Cool cool! Yeah I think we should use the "current" clear color approach. Eventually for "environmental" things I think we might need something that plays nicer with scenes, but for now I think we should go with the simple solution that works.

> I'll also work on "SubApp ZST labels instead of integers", but I could use some guidance on where to put the implementation: #2629.

Sounds good!

**Comment by superdump** [comment]
MSAA PR updated: https://github.com/bevyengine/bevy/pull/2541
Added support for configurable shadow map sizes here: https://github.com/bevyengine/bevy/pull/2700

**Comment by superdump** [comment]
Discussed on Discord a couple of weeks ago, moving the default near plane closer, from 1 unit to 0.1 units. I feel like it will be very common to consider bevy units as being real world metres, and not being able to see anything that is closer than 1m feels weird to me. :) https://github.com/bevyengine/bevy/pull/2703

**Comment by superdump** [comment]
Also as discussed on Discord, there was some confusion about the units for the `PointLight` `intensity` member. I dug into the code and figured it out that practically on `main` and before the PR it is 'luminous intensity' in lumens per steradian, which is a bit of a weird unit to figure out. Instead the Filament document had intended that it be 'luminous power' in lumens, which is what household light bulbs are rated in at the point of sale these days. Looking into pbr.wgsl there was some confusion about where a factor of 4 pi had gone - it was hidden among some other formulae but stated as luminous intensity for a point light = luminous power / 4 pi. So, I made a PR to fix this, added in the division by 4 pi, and documented the intensity value as being in units of lumens, including a table from wikipedia mapping various power ratings of household light bulbs of different types to lumen values as an aid. https://github.com/bevyengine/bevy/pull/2704

**Comment by superdump** [comment]
Should we add transparency to the to do list? I suppose as it seems to be a bit broken in 0.5, it is perhaps ok that it be broken in 0.6 as well. But otherwise I guess we would want to look into doing an opaque pass and a transparent pass and possibly implementing weighted order-independent transparency for the transparent pass.

**Comment by superdump** [comment]
Infinite reverse right-handed projection PR is now updated and fixed! \o/ https://github.com/bevyengine/bevy/pull/2543

**Comment by superdump** [comment]
Can we add a link to https://github.com/bevyengine/bevy/pulls?q=is%3Apr+is%3Aopen+base%3Apipelined-rendering to the original post? It lists the PRs against the pipelined-rendering branch of this repository.

**Comment by cart** [comment]
> So, I made a PR to fix this, added in the division by 4 pi, and documented the intensity value as being in units of lumens, 

Yup just read the section of the Filament doc you linked to. The fix in your PR looks good to me (just merged it).

> Should we add transparency to the to do list? I suppose as it seems to be a bit broken in 0.5, it is perhaps ok that it be broken in 0.6 as well. But otherwise I guess we would want to look into doing an opaque pass and a transparent pass and possibly implementing weighted order-independent transparency for the transparent pass.

I agree that ultimately we should try to do order-independent transparency, but I don't want to hold back 0.6 for it. If we can get it in before release ... awesome. But I think depth-sorting is a reasonable approach in the interim. We've gotta start pushing 0.6 out the door or it will never happen :)

> Infinite reverse right-handed projection PR is now updated and fixed!

Brilliant! I'll take a look.

> Can we add a link to https://github.com/bevyengine/bevy/pulls?q=is%3Apr+is%3Aopen+base%3Apipelined-rendering to the original post? It lists the PRs against the pipelined-rendering branch of this repository.

Good call!

**Comment by superdump** [comment]
I made a PR to add support for marking meshes should not cast or should not receive shadows: https://github.com/bevyengine/bevy/pull/2726
Also one to use a separate vertex shader for shadow passes as they only need to calculate the clip position, not the world position, UVs, nor normal as the pbr.wgsl needs. This then allows reordering the bind groups of the pbr shader to be as we would like: view, material, mesh. https://github.com/bevyengine/bevy/pull/2727

**Comment by superdump** [comment]
I naively hacked in support for glTF tangents and normal maps. Basically using the same code as we have in the current renderer, and doing whatever was needed to handle the additional vertex attribute, and so a separate pipeline layout and pipeline for when a normal map is present. But it works. And I commented out the loading of the tangent vertex attribute and normal map texture in the glTF loader and it uses the regular no normal map pipeline. I'll clean it up soon and make a PR. It's not going to be so pretty, but it's functional.

**Comment by superdump** [comment]
Here's a first shot at a PR for tangent vertex attributes and normal maps, including loading them from glTF models: https://github.com/bevyengine/bevy/pull/2741

**Comment by cart** [comment]
Just updated the task list to better represent the current state of things.

**Comment by superdump** [comment]
I've implemented a good chunk of stuff for the 'over' operator alpha blending for proper opaque and transparency, as well as properly supporting alpha masking with a cutoff, according to the glTF specs. However, to finish it off fully we need support for the double-sided property which when true means back face culling is disabled. This needs pipeline specialisation. As part of this work, it also made sense to do a depth pre-pass because alpha masking needs a special pass for depth which has a fragment shader and samples the base colour texture to identify if the fragment is opaque or not. So I did that too.

I've also been working on visibility, or at least frustum culling. I have the basic culling working but I need to put some more thought into how to handle the lists of visible entities on views neatly and efficiently. I'll need to finish this off on top of @cart 's `custom-shaders` branch as that moves to storing dynamic uniform offsets on entities, which is kind of necessary for flexibility reasons (everything just iterates over an enumerated Vec of extracted meshes and the dynamic uniform offsets are tied to this enumeration so it's a bit rigid for now.)

**Comment by superdump** [comment]
Visibility culling has progressed nicely today on top of custom-shaders now known as modular-rendering. I’ve got camera and light frustum culling working in simulation world systems, I’ve copied over RenderLayers and VisibleEntities, as well as Visible renamed to Visibility and a similar ComputedVisibility component.

Visibility is used for user control of whether an entity is visible or not. The culling system checks that and the render layers and then does frustum culling, which if the entity is not culled leads to the entity being added to that view’s visible entities as well as the entity’s computed visibility being set to true.

Then when extracting, we only extract entities that have computed visibility true as that means they are visible to one or more frusta. Visible entities are also extracted onto the view entity. Visible entities are iterated over for queuing for draws.

The NotShadowCaster component is taken into account so only casters inside the frustum are queued for shadow passes. And I did a bit of optimisation and cut down the frustum vs obb intersection test by 50%.

I think it’s close to a PR once modular-rendering is merged. :)

Future work would be to tackle calculation of the directional light shadow region and fit the frustum to it. Then implement cascade shadow maps. :)

**Comment by superdump** [comment]
Previously back when I added the tracing spans to wrap stages and systems for profiling purposes, I had tried tracing-flame which produces flamegraphs (good in theory for an aggregate view of how much time is spent in what systems, but in practice it is only proportional and not some statistical but still absolute units representation so you only see that a system took 10% not that it took N ms, which I don’t find very useful.) tracing-chrome worked well for a timeline type instrumentation. The Tracy live-visualisation application crashed a lot on macOS and I didn’t manage to fix whatever was wrong. However, since then it seems that Tracy now works quite well on macOS from some testing today and it can show a live-updating timeline, mean system execution durations and more. And you can save the captures for later offline inspection or use a headless capture tool instead of the live visualisation tool. I think I’ll make a PR against main to add it as a profiling option because it’s really good!

![](https://media.discordapp.net/attachments/886579008484016130/887731351351078932/Screenshot_2021-09-15_at_18.07.19.png)

**Comment by cart** [comment]
Just put out #2831, which significantly improves the modularity / ergonomics of render logic using "ECS driven rendering".

**Comment by superdump** [comment]
Continuing on from visibility culling, and prompted a bit by someone who was trying to render many point lights in a scene, I’ve implemented storing our point lights in a storage buffer. This touched on the need for something like UniformVec but with a BufferUsage of STORAGE instead of UNIFORM. The short version is that I’ll probably rework UniformVec at least, and at most BufferVec (which is one of the bullet points above), UniformVec and DynamicUniformVec depending on what we land on after a bit of discussion. I think I still need to remove the upper limit to the number of point lights in some places though. After I’ve resolved that, I’ll implement culling of point lights against the view frustum which will allow for many more lights to be in a larger scene though performance will be impacted if they are too densely positioned so that too many fit in the view frustum. I think this also steps toward clustered forward rendering though I need to read more about how that works exactly.

**Comment by zicklag** [comment]
I'm not sure if this is something to think about yet, but it may be worth noting that WebGL2 won't support using storage buffers in the shaders. WGPU will have a way to detect whether or not storage buffers are supported or not and allow our code to make different decisions based on that, but the renderer would still have to have different shader entrypoints and rendering paths.

**Comment by cart** [comment]
Hmm that is absolutely something we'll need to think about. I'd like the "bevy defaults" to work everywhere (including WebGL2). We have been discussing a move to storage buffers for our lighting shaders and I honestly forgot that SSBOs weren't an option on WebGL2. If WebGL2 can't support that, we will need to start thinking of ways to support that without hamstringing the rest of the renderer.

**Comment by superdump** [comment]
Here's the PR for updating `BufferVec` to use `RenderQueue` like `UniformVec`: #2847 

Also, I've been reading about [clustered forward rendering](http://www.aortiz.me/2018/12/21/CG.html) and I understand in more detail how it works now. As a brief introduction: clustered forward rendering splits the view frustum into subfrusta (a.k.a. clusters or froxels), and makes a per cluster list of point lights that are relevant to that cluster. This allows fragments falling into a cluster to only render the lights that have any significant contribution to the shading of the fragment, rather than using all lights in the scene. This in turn allows many more lights in the scene in an efficient way.

In the article, they propose having a global array of lights, an array of cluster light index lists which are indices into the global array of lights where groups of (as in consecutive runs of, though the groups are not identifiable with only this one array, keep reading and you'll understand) of indices are those relevant to an individual cluster, and an array containing, per cluster, an index into the cluster light index lists array, and a number of lights affecting that cluster. The question I was then considering was - (how) will this work with the clustered forward algorithm as proposed in the article linked above?

GpuPointLights is currently a total of 116 bytes but I think it should be 128 bytes when aligned. Given a maximum uniform binding size of 16384 bytes, that means we can have a total of 128 point lights in one uniform buffer. That's the global array of lights. Given there are 128 point lights that could be relevant to a cluster, we need log2(128) = 7 bits per index. For convenience we could think of these as u8 packed in fours into u32, which gives us 16384 indices in 4096 u32s. Using a u8 should make the masking and shifting quite nice and fast too. That's the cluster light index lists. The article proposes 16x9x24 clusters in x, y, z, respectively. That is 3456 clusters, and per cluster we need one index into the cluster light index lists array, and a count of the number of lights in the cluster. The index needs to address 16384 values which is 14 bits. The count can be no more than the 128 point lights, which is 7 bits. So we need at most 21 bits, which we can say fits in a u32 for convenience and then within 16384 bytes, we can fit up to 4096 u32 values which is more than the 3456 clusters.

So in (pseudo?) wgsl it could look like this:
```rust
[[block]]
struct PointLights {
    data: array<PointLight, 128>;
};

[[block]]
struct ClusterLightIndexLists {
    data: array<u32, 4096>;
};

[[block]]
struct ClusterOffsetsAndLengths {
    data: array<u32, 4096>;
};

[[group(0), binding(6)]]
var point_lights: PointLights;
[[group(0), binding(7)]]
var cluster_light_index_lists: ClusterLightIndexLists;
[[group(0), binding(8)]]
var cluster_offsets_and_lengths: ClusterOffsetsAndLengths;
```

...I uh think that instead of just culling lights to the frustum, I may try implementing clustered forward rendering this week using only uniform buffers (for the sake of supporting WebGL2), as the number of clusters and lights feels quite reasonable.

Possible stumbling blocks could be the performance of actually testing all the lights against all the clusters to generate the index lists, though it's highly parallelisable work, and the gathering of the results in the end is cheap. If that proves to be too slow, a different cluster configuration (the 16x9x24) could be tested.

When we get a shader preprocessor we could support using storage buffers for much larger and dynamic array sizes, and compute shaders for much better performance as well as the ability to leverage the depth prepass (that will come in the transparency PR as previously mentioned) to identify the active clusters (basically look at the depths in the depth buffer for each fragment to identify which cluster they fall into) though that would have to also support transparent meshes... they would at least have to be in clusters closer to the camera.

**Comment by superdump** [comment]
I've rebased my depth prepass / alpha modes branch on top of my visibility culling branch. The Amazon Lumberyard Bistro scene (which hacked in lights and broken transparency) before:
![Screenshot 2021-09-16 at 14 19 18](https://user-images.githubusercontent.com/302146/134373490-998df66b-6367-4990-934b-186cf47efb2e.png)
and after:
![Screenshot 2021-09-22 at 09 15 05](https://user-images.githubusercontent.com/302146/134373587-c231dc39-0f84-480c-8fae-be0e2370fc00.png)

I also made a PR to fix the panics when resizing the window or when trying to go fullscreen: <https://github.com/bevyengine/bevy/pull/2858>

**Comment by superdump** [comment]
As #2831 is now merged, I have opened a pull request for frustum culling: https://github.com/bevyengine/bevy/pull/2861

**Comment by cart** [comment]
Just merged the modular rendering pr. I also updated the checklist in the OP. I'll be working on "pipeline specialization" and "making shaders assets again" next. I'm also reviewing the various pipelined-rendering PRs that others have submitted.

**Comment by superdump** [comment]
I just got clustered-forward rendering working on macOS. 16x9x24 clusters. No shadow mapping. 128 lights. Green tint is no/few lights affecting the cluster, red tint is 16+ lights affecting the cluster, with a smoothstep between. ~55 fps on a MacBook Pro 16 with i9 9980HK and Radeon Pro 5500M (~mobile GTX 1050 Ti). This is only using uniform buffers and no compute so should work fine on WebGL2. :)
https://user-images.githubusercontent.com/302146/135138603-12bcd1ef-b636-453b-84ad-95b00d76aed6.mp4

**Comment by superdump** [comment]
Here’s the fixed version by the way: https://twitter.com/swainrob/status/1443263405457133580?s=21 Still ironing out some more basic bugs that didn’t have visual impact. Then I’ll do a little profiling on it. Maybe it’s worth optimising the light allocation either with not naively looping over all clusters vs all lights, but iterating only the clusters that a light would affect given its position and radius, or even possibly leveraging a Bounding Volume Hierarchy of the lights and casting the cluster AABBs against it. I haven’t seen what takes time yet.

Note as well that shadow mapping falls over fast with all the culling of meshes to lights, and 6 passes per point light. We’ve discussed a few ideas how to deal with that like only updating a fixed number of shadow maps per frame, and scoring the maps according to things like their distance from the camera (further away is lower priority), and frames since last update (older is higher priority.) I thought that one would want to first identify the lights that had moved and the lights that had geometry within their range move (i.e. either the light or occluders in the light’s range moved) but that is more complicated and difficult to implement with good performance.

However, at this point I don’t want to go too deep on any one thing because there are too many low-hanging fruit so I think I’ll revisit the other path I mentioned - automatically fitting the directional light orthographic projection to the view frustum and relevant geometry and maybe then implement cascade shadow maps so shadows in outdoor scenes are decent. :)

**Comment by Tarnadas** [comment]
Hey,
I just found out that wgpu now has initial WebGL support: https://github.com/gfx-rs/wgpu/blob/master/CHANGELOG.md#wgpu-011-2021-10-07
How does that affect Bevy? Would I be able to run Bevy on WASM without the external WebGL2 plugin?

**Comment by zicklag** [comment]
@Tarnadas That's the plan! I don't think it's working yet, but hopefully Bevy will pretty much "just work" once we get everything sorted out.

Currently most of the WGPU examples are working on WebGL, and I did get at least the 2D Bevy examples to work with OpenGL, which shares 90% of the code with WebGL, so at this point we've mostly just got to try stuff, and fix any issues with OpenGL/WebGL that we run into, because there are almost surely going to be some bugs in the backend that we run into.

**Comment by billyb2** [comment]
Hey just out of curiosity, I have a few questions:

So I'm pretty sure that when this PR gets merged, Bevy 0.6 will be released(!) But is there a timescale on when that will be?
Also, should I report rendering buts on this issue? I managed to find one though I'm not sure if this is the place, since I know the whole task is very experimental.

**Comment by cart** [comment]
Sprite Batching and Normal Maps are now merged!

**Comment by superdump** [comment]
Frustum culling lined up.

**Comment by superdump** [comment]
* https://github.com/bevyengine/bevy/pull/2861
* https://github.com/bevyengine/bevy/pull/3072 (based on top of frustum culling)

**Comment by mnmaita** [comment]
@cart is it fine if I try to tackle the `run_sub_graph` API one? I might push something later today or during the weekend.

**Comment by cart** [comment]
Absolutely. Thanks!

**Comment by superdump** [comment]
Lined up https://github.com/bevyengine/bevy/pull/3126 and going to get clustered-forward rendering in shape next. I rebased it once but something broke and gave everything a green tint. I'm suspecting a change in bindings somewhere.

**Comment by superdump** [comment]
I'm looking at my clustered forward rendering branch and trying to get it in shape to make a PR.

**Comment by superdump** [comment]
* https://github.com/bevyengine/bevy/pull/3153

I know @cart wants to keep things moving so I've made the PR without writing the thorough description that I would like to, and without adding in comments with reference materials. I'll back-fill with that information as my top priority with my time.

**Comment by superdump** [comment]
I haven't specifically sought out problem cases, but the updated crevice `std140_size_static` seems to be correct for the main 2D/3D cases.
* https://github.com/bevyengine/bevy/pull/3168

**Comment by cart** [comment]
I just created the pr to merge the `pipelined-rendering` work into main. Read the PR for more details. Still a bit more work to do, but we're close!

**Comment by superdump** [comment]
> I just created the pr to merge the `pipelined-rendering` work into `main`. Read the PR for more details. Still a bit more work to do, but we're close!

Sqweeee!!

**Comment by lukors** [comment]
Here's a link to the PR: https://github.com/bevyengine/bevy/pull/3175

**Comment by cart** [comment]
#3175 has been merged! All renderer development will happen on `main` from now on!

**Comment by colepoirier** [comment]
@cart heads up the crevice issue has been closed but the task has not been marked as completed

**Comment by cart** [comment]
Just merged clustered forward rendering! We're getting close :)

**Comment by cart** [comment]
@colepoirier yeah I think crevice might be good now. I checked off the subtask because the crevice repo closed it out, but we should still do one last validation pass before considering crevice "done".

**Comment by aloucks** [comment]
The bevy_render2 pipelined renderer seems to be tightly coupled to wgpu. Is this temporary to get things moving along, or is the intent for the new renderer to only integrate with wgpu?

**Comment by cart** [comment]
This was an intentional choice with [a lot of discussion behind it](https://github.com/bevyengine/bevy/discussions/2265). Wgpu is already a "generic gpu abstraction" and makes exactly the tradeoffs we want for a user-facing gpu api. Creating our own would literally be mirroring wgpu for zero added benefit (we did this in the old renderer and it brought nothing but pain and increased complexity). Instead, we've chosen to let wgpu be the  "generic bevy gpu api". Feedback on this design has been very positive so far (from Bevy developers building renderer features). Additionally, the wgpu team has proven to be a responsive partner. Read the link above (and the discussions that followed), for examples of how they adjusted project structure and licensing as a direct response to our feedback. The result is a much simpler API that I'm comfortable relying on (and maintaining ... if it ever comes to that).

**Comment by cart** [comment]
The `bevy_ui` port is merged! New plan: I'm going to focus on fixing the remaining show stopping bugs (such as #3043), then I'll immediately swap in the new renderer and remove the old one. My "actual pipelining" work will take a few days to finish, review, and merge. I'd rather get the new renderer in users' hands asap so we can identity and fix issues as early as possible.

**Comment by cart** [comment]
The old render has been replaced by the new renderer #3312. Give it a try and report any issues you encounter!

**Comment by kirawi** [comment]
What's the current performance looking like on the benchmark?

**Comment by Tarnadas** [comment]
Hey,
I'm trying to make this work with WASM, but I failed because wgpu seems to add ash as a dependency, which cannot be compiled on WASM.
ash is a library for Vulkan, so it should not be included.

```
❯ cargo tree -i -p ash --edges features --target wasm32-unknown-unknown --no-default-features
ash v0.34.0+1.2.203
├── ash feature "debug"
│   └── wgpu-hal v0.12.1
│       ├── wgpu-hal feature "default"
│       │   └── wgpu-core v0.12.1
│       │       ├── wgpu-core feature "default"
│       │       │   └── wgpu v0.12.0
```

but if I look into Cargo.toml from wgpu-hal, I see this:
```
[features]
default = []
metal = ["naga/msl-out", "block", "foreign-types"]
vulkan = ["naga/spv-out", "ash", "gpu-alloc", "gpu-descriptor", "libloading", "inplace_it"]

[dependencies]
ash = { version = "0.34", optional = true }
```

Someone suggested to bump the Rust edition to 2021, which partially fixed the problem for me.
I've been trying to bump bevy-webgl2 to use latest bevy main branch (I know it doesn't make sense. I just wanted it to compile, but it failed while compiling the dependencies, but the Rust edition fixed it). I then tried to update bevy-rapier to use latest bevy main branch, changed the Rust edition to 2021, but ash is still included as a dependency for no apparent reason.

This bug cannot be reproduced with the examples from the repository btw.

Link to Discord discussion: https://discord.com/channels/691052431525675048/750833140746158140/923451977181044746
